### PR TITLE
#2347 - #2606 reordering key individuals

### LINKED ISSUE #2347: Issuer ETO forms - bug when deleting Key individuals
When you delete a position in the key individuals form after 1 click on"x" the form reloads back to dashboard. There is no way to delete more positions in the form without going to dashboard first
image
.

### PR DESCRIPTION
**Pull request by AlexeyTuboltsev**
this PR 
- adds possibility to move keyIndividuals up and down
- fixes #2347

### CODE REVIEW

**Comment by AlexeyTuboltsev** [comment]
FOR QA: please test it from #2347 too

**Comment by AlexeyTuboltsev** [comment]
the .svg arrows are not done quite right yet, I'll swap them on Monday. They don't affect neither code nor css, so you can review now ;))

File: app/utils/generateKey.ts [hunk 1]
```diff
@@ -0,0 +1,37 @@
+// code taken from https://github.com/zzarcon/react-smart-key.
```
Comment (Moejoe90) [inline app/utils/generateKey.ts]: Why is this added it's not used anywhere in this PR?

File: app/components/shared/SocialProfilesEditor.tsx [hunk 2]
```diff
@@ -232,4 +258,4 @@ class SocialProfilesEditorLayout extends React.Component<IProps & TFormikConnect
   }
 }
 
-export const SocialProfilesEditor = connect<IProps, any>(SocialProfilesEditorLayout);
+export const SocialProfilesEditor = connect<IProps>(SocialProfilesEditorLayout);
```
Comment (Moejoe90) [inline app/components/shared/SocialProfilesEditor.tsx]: Taking of `any` wouldn't solve the problem as its typed now by default any 
have you tried `connect<IProps,VALUES>` based on the component up it looks that values = `IProps & TFormikConnect`
Check
https://jaredpalmer.com/formik/docs/api/connect#type-signature

File: app/components/shared/SocialProfilesEditor.tsx [hunk 3]
```diff
@@ -173,24 +175,48 @@ interface IState {
...
+    const {
+      formik: { values },
+      name,
+    } = this.props;
+    const {
+      formik: { values: oldValues },
+    } = prevProps;
+
+    if (!isEqual(get(oldValues, name), get(values, name))) {
+      this.setSelectedFields(values, name);
```
Comment (Moejoe90) [inline app/components/shared/SocialProfilesEditor.tsx]: I don't think its a good idea to call `this.setSelectedFields` from here 
as it has `setState` 
https://github.com/Neufund/platform-frontend/pull/2675/files?file-filters%5B%5D=.scss&file-filters%5B%5D=.tsx#diff-c733dbc8225e5508acd8dfd1d26b3e76R197

Is there no way out of this shouldn't formik update all values automatically 
Lets discuss this

File: app/components/eto/registration/pages/KeyIndividuals.tsx [hunk 4]
```diff
@@ -50,68 +52,100 @@ interface IIndividual {
...
+          </legend>
+          <SocialProfilesEditor
+            profiles={SOCIAL_PROFILES_PERSON}
+            name={`${member}.socialChannels`}
+          />
+        </fieldset>
+      </FormHighlightGroup>
+      <div>
+        <ButtonIcon
+          onClick={() => swap(index, index - 1)}
```
Comment (desfero) [inline app/components/eto/registration/pages/KeyIndividuals.tsx]: can you please add `alt` prop to `ButtonIcon` and pass it back to `InlineIcon`? This is bad for accessibility to have icons without any text so we can add `Move Up` and `Move Down` alternative (`alt`) text

File: app/components/eto/registration/pages/KeyIndividuals.tsx [hunk 5]
```diff
@@ -133,6 +167,17 @@ class KeyIndividualsGroupLayout extends React.Component<IKeyIndividualsGroup & T
...
 
+  shouldComponentUpdate(nextProps: IKeyIndividualsGroup & TFormikConnect): boolean {
+    const {
+      formik: { values },
+      name,
+    } = this.props;
+    const {
+      formik: { values: newValues },
+    } = nextProps;
+    return !isEqual(get(newValues, name), get(values, name));
```
Comment (desfero) [inline app/components/eto/registration/pages/KeyIndividuals.tsx]: For some reason this will not help, even when you always put `false` here it will still rerender everything. Just turn on react highlight updates feature to see that there are no benefits here

File: app/components/shared/SocialProfilesEditor.tsx [hunk 6]
```diff
@@ -1,5 +1,7 @@
 import * as cn from "classnames";
 import { connect, FieldArray, getIn } from "formik";
+import { isEqual } from "lodash/fp";
+import get from "lodash/get";
```
Comment (desfero) [inline app/components/shared/SocialProfilesEditor.tsx]: move `get` to `lodash/fp` please

File: app/components/shared/SocialProfilesEditor.tsx [hunk 7]
```diff
@@ -123,7 +125,7 @@ const SocialMediaTags: React.FunctionComponent<{
           <button
             data-test-id={`social-profiles.profile-button.${name}`}
             onClick={() => onClick(index)}
-            className={cn(Boolean(selectedFields[index]) && "is-selected", styles.tab)}
+            className={cn({ "is-selected": Boolean(selectedFields[index]) }, styles.tab)}
```
Comment (desfero) [inline app/components/shared/SocialProfilesEditor.tsx]: 🎖

File: app/components/shared/SocialProfilesEditor.tsx [hunk 8]
```diff
@@ -176,29 +178,54 @@ interface IState {
...
+  setSelectedFields = (values: IProps & TFormikConnect, name: string) => {
+    const {
+      profiles,
+      formik: { setFieldValue },
+    } = this.props;
     const socialMediaValues: TSocialChannelsType = getIn(values, name) || [];
     const selectedFields: boolean[] = [];
 
-    profiles.forEach((profile, index) => {
+    const profilesWithUserData = profiles.map((profile, index) => {
```
Comment (desfero) [inline app/components/shared/SocialProfilesEditor.tsx]: 🎖

File: app/components/shared/SocialProfilesEditor.tsx [hunk 9]
```diff
@@ -176,29 +178,54 @@ interface IState {
...
       const previousLink = socialMediaValues.find(v => v.type === profile.name);
       const value: string = previousLink ? previousLink.url || "" : "";
-      setFieldValue(`${name}.${index}`, { type: profile.name, url: value });
       selectedFields[index] = profile.preSelected ? true : !!value;
```
Comment (desfero) [inline app/components/shared/SocialProfilesEditor.tsx]: we should also refactor this, as doing side effects is not acceptable inside `map`

File: app/components/shared/SocialProfilesEditor.tsx [hunk 10]
```diff
@@ -176,29 +178,54 @@ interface IState {
...
+    const {
+      formik: { values },
+      name,
+    } = this.props;
+    this.setSelectedFields(values, name);
   }
 
-  toggleProfileVisibility = (index: number): void => {
-    const { selectedFields } = this.state;
+  shouldComponentUpdate(nextProps: IProps & TFormikConnect, nextState: IState): boolean {
```
Comment (desfero) [inline app/components/shared/SocialProfilesEditor.tsx]: save with KeyIndividuals

File: app/components/shared/SocialProfilesEditor.tsx [hunk 11]
```diff
@@ -176,29 +178,54 @@ interface IState {
...
+    return !isEqual(get(newValues, name), get(values, name)) || !isEqual(this.state, nextState);
+  }
 
-    this.setState({ selectedFields });
+  toggleProfileVisibility = (index: number): void => {
+    const {
+      formik: { setFieldValue },
+      name,
+    } = this.props;
+    const newSelectedFields = [...this.state.selectedFields];
```
Comment (desfero) [inline app/components/shared/SocialProfilesEditor.tsx]: hm, can you change `selectedFields` to ReadonlyArray? This is really bad to modify array and then push it to setState. Thanks for fixing it

Comment (desfero) [reply app/components/shared/SocialProfilesEditor.tsx]: btw, we should use function as a `setState` first arg because of `https://reactjs.org/docs/state-and-lifecycle.html#state-updates-may-be-asynchronous`

File: app/components/shared/forms/FormSection.module.scss [hunk 12]
```diff
@@ -1,24 +1,19 @@
 .form-section {
-  @include make-row;
+  display: grid;
+  grid-template: auto auto / auto;
+  grid-column-gap: 2rem;
+  width: 100%;
```
Comment (desfero) [inline app/components/shared/forms/FormSection.module.scss]: do we need `width: 100%`, how `display: grid` affect `width`? I always thought it's the same for container as `display: block`

File: app/components/shared/forms/FormSection.tsx [hunk 13]
```diff
@@ -12,7 +12,7 @@ export const FormSection: React.FunctionComponent<IProps> = ({ title, children }
   return (
     <section className={styles.formSection}>
       <h3 className={styles.title}>{title}</h3>
-      <div className={styles.content}>{children}</div>
+      <div>{children}</div>
```
Comment (desfero) [inline app/components/shared/forms/FormSection.tsx]: hm, that's the reason why i don't like fully using display grid inside css as then in react you will have a lot of plain `div` without any class. And this can be misunderstood by some of the developers and removed in future

File: app/components/eto/registration/pages/KeyIndividuals.tsx [hunk 14]
```diff
@@ -50,68 +52,100 @@ interface IIndividual {
...
+  image: string;
+  key: string;
+}
+
 const getBlankMember = () => ({
   name: "",
   role: "",
   description: "",
   image: "",
+  key: Math.random().toString(),
```
Comment (Moejoe90) [inline app/components/eto/registration/pages/KeyIndividuals.tsx]: Math.random() doesn't guarantee the uniqueness of generated values

The probability is indeed low but there. This will introduce an extremely hard to find bug in the future

File: app/components/eto/registration/pages/KeyIndividuals.tsx [hunk 15]
```diff
@@ -250,15 +299,56 @@ const EtoRegistrationKeyIndividuals = compose<React.FunctionComponent>(
...
   }),
 )(EtoRegistrationKeyIndividualsComponent);
 
 export { EtoRegistrationKeyIndividualsComponent, EtoRegistrationKeyIndividuals };
+
+const generateKeys = () => (data: { members: IMemberData[] }) => {
+  const members =
+    data.members &&
+    data.members.map(member => {
+      member.key = Math.random().toString();
```
Comment (Moejoe90) [inline app/components/eto/registration/pages/KeyIndividuals.tsx]: See comments up

File: app/utils/generateKey.ts (hunk 1 above)
Comment (Moejoe90) [inline app/utils/generateKey.ts]: Is this file used in the PR?

File: app/components/eto/registration/pages/KeyIndividuals.tsx [hunk 16]
```diff
@@ -250,15 +299,56 @@ const EtoRegistrationKeyIndividuals = compose<React.FunctionComponent>(
     }),
     dispatchToProps: dispatch => ({
       saveData: (data: TPartialCompanyEtoData) => {
-        dispatch(actions.etoFlow.saveDataStart({ companyData: data, etoData: {} }));
+        const convertedData = convert(data, fromFormState);
```
Comment (Moejoe90) [inline app/components/eto/registration/pages/KeyIndividuals.tsx]: What is the exact reason for using `convert` is it just to get rid of the additional key added? 
If that's the case then I suggest creating a utility function that utilizes `convert` is really not clear that the only reason we are doing convert is to get rid of keys

File: app/components/eto/registration/pages/KeyIndividuals.tsx [hunk 17]
```diff
@@ -250,15 +299,56 @@ const EtoRegistrationKeyIndividuals = compose<React.FunctionComponent>(
...
     validationSchema: EtoKeyIndividualsType.toYup(),
-    mapPropsToValues: props => props.stateValues,
+    mapPropsToValues: props => convert(props.stateValues, toFormState),
     handleSubmit: (values, props) => props.props.saveData(values),
   }),
 )(EtoRegistrationKeyIndividualsComponent);
 
 export { EtoRegistrationKeyIndividualsComponent, EtoRegistrationKeyIndividuals };
+
+const generateKeys = () => (data: { members: IMemberData[] }) => {
```
Comment (Moejoe90) [inline app/components/eto/registration/pages/KeyIndividuals.tsx]: Both `generateKeys`,`removeKeys` look like utility functions that shouldn't stay here IMO

File: app/components/shared/SocialProfilesEditor.tsx (hunk 3 above)
Comment (AlexeyTuboltsev) [reply app/components/shared/SocialProfilesEditor.tsx]: solved

File: app/components/eto/registration/pages/KeyIndividuals.module.scss [hunk 18]
```diff
@@ -6,6 +16,7 @@
 
 .add-button {
   padding-top: 0 !important; // sass-lint:disable-line no-important
+  padding-left: 0 !important; // sass-lint:disable-line no-important
```
[the remark this answers is not in the document]
Comment (AlexeyTuboltsev) [reply app/components/eto/registration/pages/KeyIndividuals.module.scss]: yes, I agree. I don't like this either

File: app/components/eto/registration/pages/KeyIndividuals.module.scss [hunk 19]
```diff
@@ -1,3 +1,13 @@
+.section-wrapper {
+  max-width: initial;
```
[the remark this answers is not in the document]
Comment (AlexeyTuboltsev) [reply app/components/eto/registration/pages/KeyIndividuals.module.scss]: I have to override the max-width=730px in app/components/eto/registration/Shared.module.scss .
I don't want to make any big refactoring here now. It's out of scope of this pr, easier to override it

File: app/components/shared/SocialProfilesEditor.tsx (hunk 11 above)
Comment (AlexeyTuboltsev) [reply app/components/shared/SocialProfilesEditor.tsx]: but we don't derive this state from previous state, do we?

File: app/components/shared/SocialProfilesEditor.tsx (hunk 2 above)
Comment (AlexeyTuboltsev) [reply app/components/shared/SocialProfilesEditor.tsx]: sure you're right, fixed

File: app/components/shared/forms/FormSection.module.scss (hunk 12 above)
Comment (AlexeyTuboltsev) [reply app/components/shared/forms/FormSection.module.scss]: thank you, this one is a wip leftover too

File: app/components/shared/forms/FormSection.tsx (hunk 13 above)
Comment (AlexeyTuboltsev) [reply app/components/shared/forms/FormSection.tsx]: well they'll see immediately why they shouldn't 
i see your point but I don't know what to do here

File: app/utils/generateKey.ts (hunk 1 above)
Comment (AlexeyTuboltsev) [reply app/utils/generateKey.ts]: good catch, this is a wip leftover, sorry for this

File: app/components/eto/registration/pages/KeyIndividuals.tsx (hunk 14 above)
Comment (Moejoe90) [reply app/components/eto/registration/pages/KeyIndividuals.tsx]: How do you know that the probability is zero? 

for sure we shouldn't rely on non-deterministic values 
Does key have size limitations? 
If not then why not hash `Math.random().toString()` with `UTC.timestap` we already have hashing functions in the bundle.

File: app/components/eto/registration/pages/KeyIndividuals.module.scss [hunk 20]
```diff
@@ -1,3 +1,13 @@
...
+}
+
+.wrapper {
```
[the remark this answers is not in the document]
Comment (desfero) [reply app/components/eto/registration/pages/KeyIndividuals.module.scss]: ok, let's then open a ticket

File: app/components/shared/SocialProfilesEditor.tsx (hunk 11 above)
Comment (desfero) [reply app/components/shared/SocialProfilesEditor.tsx]: but we do `this.state` a lot in `toggleProfileVisibility`. Or maybe i don't understand something

File: app/components/shared/forms/FormSection.tsx (hunk 13 above)
Comment (desfero) [reply app/components/shared/forms/FormSection.tsx]: they will not see it if `children` provides single element. So the problem can appear after refactoring of `children` for e.g. to use `<>` then it will by committed to DOM as a couple of elements and this can be really hard to catch later
